osrparse/replay.py

```diff
@@ -173,22 +104,230 @@ def _parse_replay_id(self, replay_data):
         # replay_ids, since they were likely manually migrated at some point
         # after the switch to long took place.
         # See:
-        # https://github.com/ppy/osu/blob/84e1ff79a0736aa6c7a44804b585ab1c54a84399/
-        # osu.Game/Scoring/Legacy/LegacyScoreDecoder.cs#L78-L81
+        # https://github.com/ppy/osu/blob/84e1ff79a0736aa6c7a44804b585ab1c54a843
+        # 99/osu.Game/Scoring/Legacy/LegacyScoreDecoder.cs#L78-L81
+        try:
+            replay_id = self.unpack_once("q")
         except struct.error:
-            format_specifier = "<l"
-            replay_id = struct.unpack_from(format_specifier, replay_data, self.offset)
-        self.replay_id = replay_id[0]
+            replay_id = self.unpack_once("l")
+        return replay_id
 
-    def dump(self, file=None):
-        dumped = dump_replay(self)
+    def unpack(self):
+        mode = GameMode(self.unpack_once("b"))
+        game_version = self.unpack_once("i")
+        beatmap_hash = self.unpack_string()
+        username = self.unpack_string()
+        replay_hash = self.unpack_string()
+        count_300 = self.unpack_once("h")
+        count_100 = self.unpack_once("h")
+        count_50 = self.unpack_once("h")
+        count_geki = self.unpack_once("h")
+        count_katu = self.unpack_once("h")
+        count_miss = self.unpack_once("h")
+        score = self.unpack_once("i")
+        max_combo = self.unpack_once("h")
+        perfect = self.unpack_once("?")
+        mods = Mod(self.unpack_once("i"))
+        life_bar_graph = self.unpack_string()
+        timestamp = self.unpack_timestamp()
+        play_data = self.unpack_play_data(mode)
+        replay_id = self.unpack_replay_id()
 
-        if not file:
-            return dumped
+        rng_seed = None
+        if play_data[-1].time_delta == -12345:
+            rng_seed = play_data[-1]
+            del play_data[-1]
 
-        # allow either a live file object, or a path to a file
-        if isinstance(file, TextIOWrapper):
-            file.write(dumped)
-        else:
-            with open(file, "wb") as f:
-                f.write(dumped)
+        return Replay(mode, game_version, beatmap_hash, username,
+            replay_hash, count_300, count_100, count_50, count_geki, count_katu,
+            count_miss, score, max_combo, perfect, mods, life_bar_graph,
+            timestamp, play_data, replay_id, rng_seed)
+
+
+
+class Replay:
```

Comment:
you can use a dataclass for the `Replay` class too
dataclasses support additional methods, with `self` and all of that things
and since there's no type changing stuff in the class, it's safe to change it into a dataclass

Reply:
I did consider this but decided against it because `Replay` seemed too "complicated" - I prefer to keep dataclasses for simple classes that do either nothing or very little more than hold data. That said, it would save us from creating dunder hash str and repr so it's probably worthwhile here.

Reply:
after thinking about it, it would be kinda useful but at the same time it's not much needed
would it actually benefit the class itself that much? at least it doesn't change the way for the consumer, so we can leave it as it is now